### PR TITLE
Zoutvat ws

### PR DESCRIPTION
**Pull request by Topvennie**
Rewrite of Loxsi

- Split main.py into multiple files
- Add a websocket connection between Loxsi and Telraam
- Add documentation

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,13 +1,10 @@
...
 WORKDIR /loxsi
 
-COPY pyproject.toml poetry.lock ./
-RUN poetry install --only main
+COPY requirements.txt .
+RUN pip install --no-cache-dir -r requirements.txt
 
 COPY . .
 
-CMD [ "pypy3", "-m", "uvicorn", "--host", "0.0.0.0", "--port", "80", "main:app" ]
```
Comment (FKD13) [inline Dockerfile]: - I still need to do a proper review of the code.

What is the reason to move away from pypy and poetry?
- pypy gives us some imploved performance
- poetry made it easy to manage the few dependencies we have and separate main from development dependencies

File: main.py [hunk 2]
```diff
@@ -1,190 +1,47 @@
...
-                    lap_source['id']: LapSource(**lap_source) for lap_source in lap_sources
-                }
 
-                laps: List[Lap] = [
-                    Lap(team=teams_by_id[lap['teamId']], lap_source=lap_sources_by_id[lap['lapSourceId']], **lap) for
-                    lap in laps
-                ]
+# Start background tasks when the app starts
+@asynccontextmanager
+async def lifespan(app: FastAPI):
```
Comment (FKD13) [inline main.py]:
```suggestion
async def lifespan(_app: FastAPI):
```

File: api.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+            status_code=HTTP_202_ACCEPTED,
+            dependencies=[Depends(self._admin_auth)],
+        )
+        async def _delete_freeze():
+            self._settings.freeze = None
+            self._settings.persist()
+            await self._admin_publisher.publish("freeze", None)
+
+        @self.get("/admin", response_class=HTMLResponse)
+        async def _admin(request: Request, _=self._admin_auth):
```
Comment (FKD13) [inline api.py]: Add this as a dependecy in the self.get

**Review by FKD13 — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 1 above)
Comment (Topvennie) [reply Dockerfile]: - The performance gains from pypy is negligible compared to python 3.12.2
- Poetry got added in [b8e2990](https://github.com/12urenloop/Loxsi/pull/13/commits/b8e29905f67c65273fc1e02c3462deaede547fdb)

File: Dockerfile [hunk 4]
```diff
@@ -10,4 +10,4 @@ RUN poetry install --only main
...
+# CMD ["uvicorn", "--host", "0.0.0.0", "--port", "80", "main:app"]
```
Comment (Topvennie) [inline Dockerfile]: Commented out code

File: config.yml [hunk 5]
```diff
@@ -1,10 +1,17 @@
 admin:
   name: admin
   password: admin
-freeze: null
-message: null
+interval:
+  feed: 3
+  fetch: 2
+  websocket: 5
+message: Je Mama
```
Comment (Topvennie) [inline config.yml:8]: :(

File: pyproject.toml [hunk 6]
```diff
@@ -1,21 +1,21 @@
 [tool.poetry]
+package-mode = false
 name = "loxsi"
-version = "0.6.9"
-description = "Data proxy for the Telraam application"
+version = "0.2.0"
```
Comment (Topvennie) [inline pyproject.toml]: Version doesn't correspond with the one specified in the app

File: src/data_publisher.py
Comment (Topvennie) [inline src/data_publisher.py:1]: Very nice comments

**Comment by FKD13** [comment]
closes #14

**Comment by Topvennie** [comment]
Very nice 👍

**Review by FKD13 — APPROVED** [review]

**Comment by FKD13** [comment]
Indeed :+1: 

Good programmers

**Pull request merged**